[RFC] Cleanups in core/imagedownloader.cpp

bstoeger opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
First a disclaimer: This was only compile-tested and some of it is non-trivial. Unfortunately, I am very busy at the moment and testing this fully is quite tricky. Notably, I don't know how to upload images to the cloud - much less how to test the fallback from hash to filename.

That being said, this is various cleanups and a leak-fix. `core/imagedownloader.cpp` contained some *very* scary code and I don't understand how this ever worked. If I read it right(?), it `delete`d a subobject(!). And it recursed from a slot called in an event-loop into the function event-loop-generating function. Wow.

But maybe I'm wrong and this was all fine. The whole Qt object-ownership and thread models are bizarre to me.
### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->

neolit123 commented on core/imagedownloader.cpp:
```diff
@@ -17,45 +17,50 @@ static QUrl cloudImageURL(const char *hash)
-		connect(&manager, SIGNAL(finished(QNetworkReply *)), this, SLOT(saveImage(QNetworkReply *)));
+		connect(&manager, &QNetworkAccessManager::finished, this,
+			[this,&success] (QNetworkReply *reply) { saveImage(reply, success); });
 		QNetworkReply *reply = manager.get(request);
```
AFAIK, we don't have lambdas anywhere in the code. probably best to keep it that way.

bstoeger replied:
Actually, I stumbled over quite a few of them, so I thought it was OK.

Here, not using a lambda would mean adding a `success` member variable. Or do you have a different idea how we can get the download status back to the caller?

neolit123 replied:
maybe @dirkhh changed our policy on lambdas and they are now accepted?

dirkhh replied:
I still hate them. I'm aware of a few that I had to agree made code more readable - and it's typically in "connect()" situations. I hope I'll find time to review things (and do 4.7.7 - this will definitely not go in before that) tomorrow.

bstoeger replied:
Actually, stateful lambdas are one of the features I really like about C++. Just think about all the inane boiler-plate that you can avoid by using a simple lambda instead of a functor.

atdotde commented:
I admit I wrote that code. To a very large degree ignorant of how things are supposed to work in C++ or Qt for that matter. Plus images in the cloud have never worked. There is no api for that on the cloud server and so those calls always fail. At the time, I hoped to come back to it at some point and implement the server side but that never happened. 
So we are left with three sources of images that should work: Those with actual local filenames. Those for which we have a hash (which can be translated to a local filename, possibly through the filename to hash dictionary [the idea is that you attach the images to your log on one computer and have the images on another computer but under a different path. Then, once Subsurface manages to actually load the image from the path, it remembers its hash and subsequently identifies the image via its hash. Plus there is a function to hash all images in a subtree that can be triggered manually to locate pictures that the log has hashes of but which are stored in different paths than at the time the log was created). Plus URLs. Once we manage to load a picture from a remote URL, we also store a local copy and access it via the hash.]).

This whole NetworkManager object business is totally mysteries to me, that code came to our source via cut and paste from the docs (at least where I was involved) plus a bit of trail and error in how to deal with the responses. I cannot claim that I had any mental model in mind when I wrote that but I think I got it to a point where it was operational. 
So any improvement by a more understanding author is very welcome as long as it does not breaks things that worked in the past.

bstoeger commented:
> This whole NetworkManager object business is totally mysteries to me

It seems crazy that we have to do async IO when we are already in a separate worker thread. Isn't there a sane synchronous network API?

Moreover, I wonder if it wouldn't simplify things if thumbnailing was done as a part of hashing? I will think about that.

bstoeger commented:
Please don't apply yet. Perhaps #8b3898f introduces a race:
Can `reply->isRunning()` return `false` before the slot has returned?

bstoeger commented:
> Perhaps #8b3898f introduces a race

This should be fixed in the current version, which replaces the busy-waiting loop by a signal.

Tested it by manipulating an .xml directly (replaced filename by URL and removed the old hash).

BTW: I wonder if the `updateHash` function could be replaced by `hashPicture`. It does the same thing, only it doesn't call `mark_divelist_changed` if the hash changed. Is that intended?

bstoeger commented:
Ok, I did some more detangling. Tested it by deleting the old hash file and adding URLs to the .xml file. Also, loaded from cloud storage. Everything seems to still work so far. I did not test the "find pictures" functionality.

I have a patch on top of this which removes `picture->hash` completely, as it is basically non-functional. It is only updated on loading, therefore we might just always use the hash maps and fill them on loading. But I'd rather get the cleanups in this PR applied first.

atdotde commented:
Loading and saving you mean? Part of the whole point is to write the hashes to the divelog file on disk.

bstoeger commented:
Yes, nothing is changed on saving. Only the in-memory `picture->hash` field is removed, because it is effectively replaced by the filename<->hash maps.

bstoeger commented:
Rebased to master and added hash-in-picture-struct-removal commit.

Deleted old hashe-file, added image from URL, everything seems to work fine.

Note that this PR exposes a problem: what to do if we find different hashes in the save-file and the hash-file? But at least now we have a distinct place where an appropriate policy can be implemented.

atdotde commented:
When running this on my log, it deadlocks. The main thread cannot obtain the lock in getHash() called from picture_exists() and hangs.

No other thread is currently doing anything with pictures, so that must be a dangling lock. Yes, just reading from the hash does not require a lock (or we could do with write only locks like we now have for the alternative plans) but the actual problem is deeper.

atdotde commented on core/qthelper.cpp:
```diff
@@ -1223,23 +1240,17 @@ void learnImages(const QDir dir, int max_recursions)
+	if (localPath.isEmpty())
+		return false;
+	QByteArray hash = hashFile(localPath);
+	QMutexLocker locker(&hashOfMutex);
```
Why do you lock here? You don't access the the QHashes here. In fact, getHash() hangs because it cannot obtain the lock itself. Removing this lock seems to make it work for me.

bstoeger replied:
That was an accident.

atdotde commented:
@neolit123 are we ready to merge?
